log/corelog/json_logger.go

```diff
@@ -0,0 +1,77 @@
+package corelog
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+)
+
+// RFC3339MicroTimeLayout ...
+const RFC3339MicroTimeLayout = "2006-01-02T15:04:05.999999Z07:00"
+
+type messageType string
+
+const (
+	logMessageType messageType = "log"
+)
+
+type logMessage struct {
+	Timestamp   string      `json:"timestamp"`
+	MessageType messageType `json:"type"`
+	Producer    Producer    `json:"producer"`
+	ProducerID  string      `json:"producer_id,omitempty"`
+	Level       Level       `json:"level"`
+	Message     string      `json:"message"`
+}
+
+type jsonLogger struct {
+	encoder *json.Encoder
+}
+
+func newJSONLogger(output io.Writer) *jsonLogger {
+	logger := jsonLogger{
+		encoder: json.NewEncoder(output),
+	}
+
+	return &logger
+}
+
+// LogMessage ...
+func (l *jsonLogger) LogMessage(message string, fields MessageFields) {
+	msg := logMessage{
+		MessageType: logMessageType,
+		Message:     message,
+		Timestamp:   fields.Timestamp,
+		Producer:    fields.Producer,
+		ProducerID:  fields.ProducerID,
+		Level:       fields.Level,
+	}
+	err := l.encoder.Encode(msg)
```

Comment:
Can we guarantee that the JSON messages do not get intermixed in the output when multiple writes are in parallel? For example when a step calls write in parallel or there are goroutines in use?

Reply:
As far as I see we don't do too much work in parallel.
Probably only in `hangDetector.Start` and `timeoutcommand.Command.Start`, as far as I see these messages can't mess up with the logs.